pkg/proxy/controller.go

```diff
@@ -0,0 +1,132 @@
+package proxy
+
+import (
+	"context"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/client-go/informers"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	informerfactory "github.com/karmada-io/karmada/pkg/generated/informers/externalversions"
+	clusterlisters "github.com/karmada-io/karmada/pkg/generated/listers/cluster/v1alpha1"
+	searchlisters "github.com/karmada-io/karmada/pkg/generated/listers/search/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/proxy/store"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/genericmanager"
+	"github.com/karmada-io/karmada/pkg/util/lifted"
+)
+
+const workKey = "key"
+
+// Controller syncs Cluster and GlobalResource.
+type Controller struct {
+	clusterLister   clusterlisters.ClusterLister
+	registryLister  searchlisters.ResourceRegistryLister
+	worker          util.AsyncWorker
+	informerManager genericmanager.MultiClusterInformerManager
+	store           *store.MultiClusterCache
+
+	// proxy
+	karmadaProxy *karmadaProxy
+	clusterProxy *clusterProxy
+	cacheProxy   *cacheProxy
+}
+
+// NewController create a controller for proxy
+func NewController(restConfig *restclient.Config, informerManager genericmanager.MultiClusterInformerManager, factory informers.SharedInformerFactory, karmadaFactory informerfactory.SharedInformerFactory) (*Controller, error) {
+	kp, err := newKarmadaProxy(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &store.MultiClusterCache{}
+
+	ctl := &Controller{
+		clusterLister:   karmadaFactory.Cluster().V1alpha1().Clusters().Lister(),
+		registryLister:  karmadaFactory.Search().V1alpha1().ResourceRegistries().Lister(),
+		informerManager: informerManager,
+		store:           s,
+		karmadaProxy:    kp,
+		clusterProxy:    newClusterProxy(factory.Core().V1().Secrets().Lister()),
+		cacheProxy:      newCacheProxy(s),
+	}
+
+	workerOptions := util.Options{
+		Name:          "proxy-controller",
+		KeyFunc:       nil,
+		ReconcileFunc: ctl.reconcile,
+	}
+	ctl.worker = util.NewAsyncWorker(workerOptions)
+
+	resourceEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(interface{}) {
+			ctl.worker.Add(workKey)
```

Comment:
I don't get why just add a `key` here, since the reconcile function is empty now, I can't see how to reconcile the key.

Reply:
I will show you the code in next pr.